Skip list

enqidu opened the pull request:
#### Description of Change
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.
I implemented skip list data structure
Contributors guide: https://github.com/TheAlgorithms/C-Plus-Plus/CONTRIBUTING.md
-->

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: <!-- Please add a one-line description for developers or pull request viewers -->
Skip List is very useful data structure for searching in sparse and big data. 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/941"><img src="https://gitpod.io/api/apps/github/pbs/github.com/enqidu/C-Plus-Plus.git/eefe46688deb4197d14a7d123135037c7888a304.svg" /></a>

enqidu commented:
Can someone tell me why is it failing?

kvedala commented:
> Can someone tell me why is it failing?

Please click on details and you'll find the descriptions for exactly why the tests are failing.
![Screenshot_20200707-064056](https://user-images.githubusercontent.com/7001608/86769571-56bc7f80-c01d-11ea-82ea-3187be83451d.png)

enqidu commented:
> > Can someone tell me why is it failing?
> 
> Please click on details and you'll find the descriptions for exactly why the tests are failing.
> ![Screenshot_20200707-064056](https://user-images.githubusercontent.com/7001608/86769571-56bc7f80-c01d-11ea-82ea-3187be83451d.png)

thanks

enqidu commented:
@kvedala I have this error "Consider using rand_r(...) instead of rand(...) for improved thread safety", but when I switch to rand_r windows tests fail (even though I include #include <cstdlib> library). Could you tell me how can I resolve this?

kvedala commented:
> @kvedala I have this error "Consider using rand_r(...) instead of rand(...) for improved thread safety", but when I switch to rand_r windows tests fail (even though I include #include library). Could you tell me how can I resolve this?

1. Much of the code uses C header files and C functions. If you want to continue using them, the code will be more appropriate for the https://github.com/TheAlgorithms/C repo. 
2. For C++, the correct way to invoke `randr` function is: `std::randr`. Moreover, before invoking the rand function, the generator needs to be initialized: `std::srand(std::time(nullptr))`. 
3. Following C++ standards should help you get closer to overcoming the errors and alerts reported by the auto-checks. 

For more info, please refer to existing code in the repo for reference. There are many programs that use the above functions and should help you learn how to best use them. 

Let me also give you a heads up about the documentation standards - we use Doxygen to autogenerate the website - https://thealgorithms.github.io/C-Plus-Plus

kvedala commented:
Please also enable GitHub Actions in your fork so that the auto tools can help auto-format the code for you

enqidu commented:
> > @kvedala I have this error "Consider using rand_r(...) instead of rand(...) for improved thread safety", but when I switch to rand_r windows tests fail (even though I include #include library). Could you tell me how can I resolve this?
> 
> 1. Much of the code uses C header files and C functions. If you want to continue using them, the code will be more appropriate for the https://github.com/TheAlgorithms/C repo.
> 2. For C++, the correct way to invoke `randr` function is: `std::randr`. Moreover, before invoking the rand function, the generator needs to be initialized: `std::srand(std::time(nullptr))`.
> 3. Following C++ standards should help you get closer to overcoming the errors and alerts reported by the auto-checks.
> 
> For more info, please refer to existing code in the repo for reference. There are many programs that use the above functions and should help you learn how to best use them.
> 
> Let me also give you a heads up about the documentation standards - we use Doxygen to autogenerate the website - https://thealgorithms.github.io/C-Plus-Plus

Thank you, I've polished code following C++ standards

kvedala commented:
> This pull request **introduces 1 alert** when merging [5068ddc](https://github.com/TheAlgorithms/C-Plus-Plus/commit/5068ddc14d093c2434edcb359fca0d3594dff86c) into [5baf1ad](https://github.com/TheAlgorithms/C-Plus-Plus/commit/5baf1ad89fee29b5b394bcfe4eafc2327613eaff) - [view on LGTM.com](https://lgtm.com/projects/g/TheAlgorithms/C-Plus-Plus/rev/pr-2c63a26284f364899876acafb254e4fa1d73deaa)
> 
> **new alerts:**
> 
> * 1 for Resource not released in destructor

☝️

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,197 @@
+
+#include <iostream>
+#include <vector>
+#include <stdio.h>
```
please remove C header files. This is a C++ code!

enqidu replied:
done

Panquesito7 replied:
If this is still used, use `cstdio`, which is a C++ library.

enqidu replied:
No, I've deleted it. Thanks for your comment! by the way are we going to merge this or is there anything that must be changed?

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,197 @@
+/**
```
Please see [`data_structures/trie_modern.cpp`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/data_structures/trie_modern.cpp) as an example for implementation and documentation standards.

kvedala replied:
☝️

enqidu replied:
Should I return inner function comments?

kvedala replied:
what does that mean?

enqidu replied:
comments inside the function

kvedala replied:
inline comments in codes are always a great idea whether it is mandatory or not 😄 

But, please, please.. check the links I have shared multiple times to format the docs. We are following doxygen standard - links are all provided multiple times above. 
Thanks.

kvedala replied:
still not done.

enqidu replied:
Hello, I was away. I corrected and add  parameters/return value comments.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,197 @@
+using std::vector;
+using std::endl;
+
+#define MAXLVL 2
```
use descriptive macro names as in [this file](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/numerical_methods/newton_raphson_method.cpp) otherwise, it is difficult for a reader to learn from the code.

enqidu replied:
ok, thank you

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,197 @@
+*/
+class Node { 
+public: 
+    int key; 
```
should some of the variables be private? if not, shouldn't `struct` be more appropriate and efficient data type?

enqidu replied:
Yeah, that's correct. I've changed it into struct

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,197 @@
+    this->key = key; 
+  
+    /*Initialization of forward vector*/
+    for (int i = 0; i < sizeof(Node*)*(level+1); i++){
```
from this definition, the loop termination condition is:
> till `i` is less than the platform's pointer size times ( the number of levels + 1). 

Does this even work? on modern machines, the `sizeof(Node*)` evaluates to 4. So, the loop will proceed 12 times when `level = MAXLVL`. Is this the intent here?

enqidu replied:
Yeah, it should've been just for (int i = 0; i < (level+1); i++). You are correct.
Thank you again

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,197 @@
+*/
+int SkipList::randomLevel() { 
+    int lvl = 0; 
+    while(static_cast<float>(std::rand())/RAND_MAX < P && lvl < MAXLVL) lvl++;  
```
if you re looking for a random number less than `MAXLVL`, wouldn't it simply be:
```cpp
rand_lvl = std::rand() % MAXLVL;
```
No loops.

enqidu replied:
It's not just random number between 0 and MAXLVL, with equal probabilities. Probability of every higher level is two times less probable because of the loop. Loop doesn't take much time as number of levels = O(log(n))

kvedala replied:
got it

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,197 @@
+void SkipList::insertElement(int key, void* value) { 
+    std::cout << "Inserting" << key << "...";
+    Node *x = header; 
+    Node *update[MAXLVL+1]; 
```
here, `update` is a *pointer to* an *array of* `Node` data type. Is this the intent?

enqidu replied:
yes

kvedala replied:
in that case line# 103 doesn't make sense and my comment there stands.

enqidu replied:
I misinterpret your first remark. update is an array of Node pointers.

kvedala replied:
as mentioned in the comment above, this line crates `update` as a pointer of pointer. That is not what you wanted to do .

kvedala replied:
after reviewing, it seems to be correct. My apologies. 
update is indeed supposed to be a an array of pointers.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,212 @@
+SkipList::~SkipList(){
+    delete header;
+    for(int i=0; i <= level; i++) { 
+        Node *node = header->forward[i]; 
```
by here, `header` variable is already deleted and invalid. the `header` variable should be deleted at the end, after the loop.

enqidu replied:
yeah correcto, deleting header should happen afterwards

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,212 @@
+
+    for(int i = level; i >= 0; i--) { 
+        while(x->forward[i] != NULL && x->forward[i]->key < key)  x = x->forward[i]; 
+        update[i] = x; 
```
this is invalid because only `update[0]` is defined.

enqidu replied:
I don't get it. size of update is (MAXLVL+1)* sizeof(Node*); on 95 line memset is used for initialization.

kvedala replied:
`memset` does not create, it simply sets the values. If out of bounds, segmentation fault will occur as shown in the screenshot. 
please ensure the code compiles and executes correctly. All reviewers and contributors are volunteers devoting their own time. Hence, a self-assessment of the code by the authors will be helpful.

enqidu replied:
Yes, I understand and I'm very glad for devotion. memset shouldn't happen to reach out of bounds as size of update array equals  (MAXLVL+1)* sizeof(Node*); which is also bounds for memset.
I devoted hours studying the code yesterday and I find it very elegant implementation.

kvedala replied:
understood - `update` is an array of pointers.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,212 @@
+  
+int main() 
+{ 
+    std::srand(time(nullptr));
```
```suggestion
    std::srand(std::time(nullptr));
```

enqidu replied:
I tried that but compliator tells me that "error: ‘time’ is not a member of ‘std’"

kvedala replied:
thats because, you have been using the C libraries and not the C++

enqidu replied:
Ok, changed into <ctime>, thank you again

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,212 @@
+    SkipList lst;
+
+    for (int j = 0; j < (1 << (MAXLVL+1)); j++){
+         int k = (std::rand()% (1 << (MAXLVL+1)  + 1));
```
this line implies:
0 <= k < 9
Nothin wrong here, just adding a note for self for review.

kvedala commented:
Based on the issues seen here, I checked the provided reference: https://iq.opengenus.org/skip-list and the proposed code is an exact replica of the code provided there. This is an infringement of the copyrights of the original author there, unless you have had a written (or via email) permission from the author to use it or if the code is released under an open-source license.

enqidu commented:
> Based on the issues seen here, I checked the provided reference: https://iq.opengenus.org/skip-list and the proposed code is an exact replica of the code provided there and no significant modifications to call the code a self-implementation. This is an infringement of the copyrights of the original author there, unless you have had a written (or via email) permission from the author to use it or if the code is released under an open-source license or the code is significantly modified from the original author.
Yes, that's correct .It's a modification of the provided code. (I deleted unnecessary node creation, added a code for destroying the class and modified randomization methods ). If copyright issue is the case than I should certainly reach out for the author. But it should also be noticed that GeeksForGeeks implementation is the also very similar code 
https://www.geeksforgeeks.org/skip-list-set-3-searching-deletion/
That's why I decided that provided code was free to use (and change).

kvedala commented:
> > Based on the issues seen here, I checked the provided reference: https://iq.opengenus.org/skip-list and the proposed code is an exact replica of the code provided there and no significant modifications to call the code a self-implementation. This is an infringement of the copyrights of the original author there, unless you have had a written (or via email) permission from the author to use it or if the code is released under an open-source license or the code is significantly modified from the original author.

> Yes, that's correct .It's a modification of the provided code. (I deleted unnecessary node creation, added a code for destroying the class and modified randomization methods ). If copyright issue is the case than I should certainly reach out for the author. But it should also be noticed that GeeksForGeeks implementation is the also very similar code
> https://www.geeksforgeeks.org/skip-list-set-3-searching-deletion/
> That's why I decided that provided code was free to use (and change).

ok, thank you for the clarification. Your note will be recorded for reference to avoid any issues in the future.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+#include <iostream>
+#include <vector>
+#include <string.h>
+#include <stdlib.h>
```
there are still some C headers in the code

enqidu replied:
Yes, they are needed for memset function

kvedala replied:
every C header has a corresponding C++ analog.
* for `string.h`, it will be `cstring` -> provides `std::memset`
* for `stdlib.h`, it will be `cstdlib` -> why this?

enqidu replied:
Oh, I didn't know that.
Including stdlib.h is just  habit. It's not needed here.

enqidu replied:
Memory issues shouldn't be the case

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+using std::endl;
+
+#define MAXLVL 2   ///< maximum level of skip list
+#define P 0.5      ///< current probability for "coin toss"
```
use a better macro name as:
```suggestion
#define PROBABILITY 0.5      ///< current probability for "coin toss"
```

enqidu replied:
corrected, thanks

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+using std::vector;
+using std::endl;
+
+#define MAXLVL 2   ///< maximum level of skip list
```
use a better macro name
```suggestion
#define MAX_LEVEL 2   ///< maximum level of skip list
```
can this value be increased? if yes, to what limit? if not, why?

enqidu replied:
Yes it can, depends on the problem. For most of the cases it's better be log(n) where n is number of elements but if I remember correctly Skip Lists are also used on sparse arrays where some everistics are need.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+};
+
+  
+Node::Node(int key, int level, void* value) { 
```
missing function documentation

enqidu replied:
done

kvedala replied:
function documentation involves description of parameters. Please review the shared links for doxygen formatting standards and also the links for example codes.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+    int key; 
+    void* value;
+    /*Forward Array*/
+    vector<Node*> forward;
```
missing class member documentations

enqidu replied:
done

kvedala replied:
Please review the documentation for classes. This is not per the doxygen standards as shared in the links.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+} 
+  
+// Class for Skip list 
+class SkipList { 
```
missing proper class documentations

kvedala replied:
Please review class documentation - member variables are not documented.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+
+
+  
+SkipList::SkipList() { 
```
missing function documentation

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+} 
+
+
+SkipList::~SkipList(){
```
missing function documentation

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+  
+  
+// Insert given key in skip list 
+void SkipList::insertElement(int key, void* value) { 
```
missing function documentation

enqidu replied:
done

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+  
+    x = x->forward[0]; 
+  
+    bool doesnt_exist = (x == NULL || x->key != key);
```
in C++, NULL is deprecated and `nullptr` should be used instead. Please update wherever being used.

enqidu replied:
alright

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+} 
+  
+// Display skip list level wise 
+void SkipList::displayList() { 
```
missing function documentation

enqidu replied:
done

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,207 @@
+    } 
+} 
+  
+int main() 
```
missing function documentation

enqidu commented:
I've corrected the code. If there is any issue I'm looking forward to it

kvedala commented:
> I've corrected the code. If there is any issue I'm looking forward to it

The comments do not conform to the suggested references. The documentation standard is [doxygen](https://www.doxygen.nl/manual/docblocks.html).

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,228 @@
+/*
```
this block is still not up to the standards. Please check the above comments, references, links etc provided to you for your convenience.

kvedala replied:
There are links, references, function documentations, sample codes provided. please check them.

Panquesito7 commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,228 @@
+#include <cstring>
+#include <ctime>   
+
+// using namespace std; 
```
Please remove this comment.

enqidu replied:
okay

Panquesito7 commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,228 @@
+*/
+struct Node { 
+    int key;
+    //pointer of value 
```
Please add a space after the comment:
```c++
    // Comment (good)
    //Comment (wrong)
```
Same in other parts of the code.

enqidu replied:
ok

Panquesito7 commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,228 @@
+
+/**
+ * Main function:
+ * Creates and inserts random 2^[number of levels] elements into the skip lists and than displays it
```
Line very long. Consider making it shorter:
```suggestion
 * Creates and inserts random 2^[number of levels]
 * elements into the skip lists and than displays it
```

kvedala commented:
@Panquesito7 let there be one reviewer at a time to avoid confusions. If you would like to continue the review, I'll wait till your approval.

enqidu commented:
@kvedala I added documentation and comments as you suggested. Waiting for merge ^_^

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,241 @@
+/**
```
use the comment structure as below:
```cpp
/**
 * @file 
 * @brief Add one line description here
 * @details 
 * This is a multi line
 * description containing links, references,
 * math equations, etc in Markdown syntax
 * @author [Name](https://github.com/handle)
 */
```

this is the structure being requested in the comments [above](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/files#r451690314)

enqidu replied:
I see. I'll change it

enqidu replied:
Done (?)

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,246 @@
+ * help of hierarchy of linked lists that connect increasingly sparse subsequences of the items
+ * 
+ * References used: GeeksForGeeks skip list code, https://iq.opengenus.org/skip-list/  PseudoCode and Code
+ * @author enqidu
```
```suggestion
 * @author [enqidu](https://github.com/enqidu)
```

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,246 @@
+ * A skip list is a data structure that is used for storing a sorted list of items with a
+ * help of hierarchy of linked lists that connect increasingly sparse subsequences of the items
+ * 
+ * References used: GeeksForGeeks skip list code, https://iq.opengenus.org/skip-list/  PseudoCode and Code
```
```suggestion
 * References used: [GeeksForGeek](https://www.geeksforgeeks.org/skip-list/), [OpenGenus](https://iq.opengenus.org/skip-list) for PseudoCode and Code
```

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,246 @@
+/**
+ * @file skip_list.cpp
+ * @brief data struccture for fast searching and insertion (O(log(n)))
```
```suggestion
 * @brief Data structure for fast searching and insertion in \f$O(\log n)\f$ time
```

enqidu commented:
Hello, @kvedala I merged commits as you told me.

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,246 @@
+    void* value;
+    // Forward Array points to the neighbour (right)
+    // nodes of the given one in all levels
+    vector<Node*> forward;
```
Class variables still not commented correctly. Please refer the following
*  [Doxygen details](https://www.doxygen.nl/manual/docblocks.html)
*  [CONTRIBUTION.md](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program)

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,246 @@
+/**
+ * SkipList class implementation with basic methods
+*/
+class SkipList { 
```
Same as above comment. Please check.

kvedala commented:
Thank you. Much has been resolved. Just the class member variables are not documented per standard.

kvedala commented:
![Screen Shot 2020-07-11 at 6 41 19 PM](https://user-images.githubusercontent.com/7001608/87235086-1e62cb80-c3a6-11ea-81cd-460b183bf6c9.png)
Note:
1. The formatter is trying to apply some fixes. Please enable "GitHub Actions" on your fork
2. Better to use `std::unique_ptr` where possible. I have disabled this as an error but is a good CPP Standard Recommendation

kvedala commented:
Doc fixes look good 👍 
Self tests are [failing](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/checks?check_run_id=862949776)

enqidu commented:
> Doc fixes look good 
> Self tests are [failing](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/checks?check_run_id=862949776)

Yeah, I have no idea why, I just changed comments

Panquesito7 commented:
https://github.com/TheAlgorithms/C-Plus-Plus/runs/862949776

kvedala commented:
> > Doc fixes look good
> > Self tests are [failing](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/checks?check_run_id=862949776)
> 
> Yeah, I have no idea why, I just changed comments

We recently added stricter checks on the code submitted (see #947)

enqidu commented:
> > > Doc fixes look good
> > > Self tests are [failing](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/checks?check_run_id=862949776)
> > 
> > 
> > Yeah, I have no idea why, I just changed comments
> 
> We recently added stricter checks on the code submitted (see #947)

You told me recently that code was okay and it was free to merge. Now you have added more and more rules. I can't work on this any longer.

enqidu commented:
> > > Doc fixes look good
> > > Self tests are [failing](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/checks?check_run_id=862949776)
> > 
> > 
> > Yeah, I have no idea why, I just changed comments
> 
> We recently added stricter checks on the code submitted (see #947)

Sorry. I can't work on this any longer.

kvedala commented:
> > > > Doc fixes look good
> > > > Self tests are [failing](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/checks?check_run_id=862949776)
> > > 
> > > 
> > > Yeah, I have no idea why, I just changed comments
> > 
> > 
> > We recently added stricter checks on the code submitted (see #947)
> 
> Sorry. I can't work on this any longer.

Thank you for your tries and contribution, I'll for it to my branch and fix it further by myself. Your commits will retain your history so that you get due credit for your work.

kvedala commented:
![image](https://user-images.githubusercontent.com/7001608/87256763-727fb580-c463-11ea-9a8c-c94cf6334a9b.png)
Is this the expected output?

enqidu commented:
> ![image](https://user-images.githubusercontent.com/7001608/87257364-69451780-c468-11ea-8ddf-f0ce0e24f4d7.png)
> 
> Is this the expected output?

Yes. Numbers are randomized so every execution will give different output but list display shows us that it's working.

enqidu commented:
> > > > > Doc fixes look good
> > > > > Self tests are [failing](https://github.com/TheAlgorithms/C-Plus-Plus/pull/941/checks?check_run_id=862949776)
> > > > 
> > > > 
> > > > Yeah, I have no idea why, I just changed comments
> > > 
> > > 
> > > We recently added stricter checks on the code submitted (see #947)
> > 
> > 
> > Sorry. I can't work on this any longer.
> 
> Thank you for your tries and contribution, I'll for it to my branch and fix it further by myself. Your commits will retain your history so that you get due credit for your work.

Okay. Good luck

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,219 @@
+        }
+    }
+
+    /*
```
```suggestion
    /**
```

kvedala commented on data_structures/skip_list.cpp:
```diff
@@ -0,0 +1,219 @@
+        lst.insertElement(k, &j);
+    }
+
+    lst.displayList();
```
```suggestion
    lst.displayList();
    
    return 0;
```
